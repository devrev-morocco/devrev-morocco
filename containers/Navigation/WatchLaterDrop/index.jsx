import React, { memo, useState, useEffect, useRef, Fragment } from 'react';
import { useMediaQuery } from '../../../hooks';
import { MenuTransition } from '../../../components';
import { mobileCheck } from '../../../utils';
import { WLCartContainerWrap, WLCart_Container, WLCartHeader } from './style';
import { useWL } from '../../../hooks';
import WlContent from './WlContent';
import NoWlContent from './NoWlContent';
import PropTypes from 'prop-types';

const WatchLaterDrop = ({
  Show,
  ShoppingCartMouseIn,
  ShoppingCartMouseLeave
}) => {
  const CartNodeRef = useRef(null);
  const [IsMobile, setIsMobile] = useState(false);

  const mediaQueryMatches = useMediaQuery('max-width', 735);
  const [StoredValue] = useWL();

  const WlLength = StoredValue?.wl.length;
  const IsWatchLater = WlLength > 0;

  useEffect(() => {
    if (mobileCheck()) {
      setIsMobile(true);
    }
  }, []);

  const OnEnter = () => {
    let dropNode = document.getElementById('watch-later-cart');
    if (dropNode) dropNode.style.display = 'block';
  };

  const OnExit = () => {
    let dropNode = document.getElementById('watch-later-cart');
    if (dropNode) dropNode.style.display = 'none';
  };

  return (
    <MenuTransition
      ref={CartNodeRef}
      Show={Show}
      onEnterFunc={OnEnter}
      onExitedFunc={OnExit}
    >
      <WLCart_Container
        ref={CartNodeRef}
        isMobileMode={mediaQueryMatches}
        id="watch-later-cart"
        onMouseEnter={ShoppingCartMouseIn}
        onMouseLeave={ShoppingCartMouseLeave}
      >
        <WLCartHeader>
          {IsWatchLater && (
            <Fragment>
              <div className="wl-tt">Watch later</div>
              <div className="wl_line"></div>
              <div className="wl-ss">{`${
                WlLength > 1 ? `${WlLength} videos` : `1 video`
              }`}</div>
            </Fragment>
          )}
        </WLCartHeader>
        <WLCartContainerWrap
          isMobileMode={mediaQueryMatches}
          isMobile={IsMobile}
        >
          {IsWatchLater ? <WlContent /> : <NoWlContent />}
        </WLCartContainerWrap>
      </WLCart_Container>
    </MenuTransition>
  );
};

WatchLaterDrop.propTypes = {
  Show: PropTypes.bool,
  ShoppingCartMouseIn: PropTypes.func,
  ShoppingCartMouseLeave: PropTypes.func
};

export default memo(WatchLaterDrop);
