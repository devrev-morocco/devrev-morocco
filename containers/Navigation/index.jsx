import React, {
  memo,
  useReducer,
  useState,
  useCallback,
  useEffect,
  useRef,
  useMemo
} from 'react';
import {
  Nav,
  NavContainer,
  NavLogo,
  MenuContainer,
  MenuMainPlaceholder,
  MenuBurgerContainer,
  Burger,
  BurgerTxt,
  RightContainer,
  WatchLaterContainer,
  ClockIconContainer,
  WLCounter
} from './styles';
import { WLMenuTriangle } from '../../styles/StyledComponents';
import Link from 'next/link';
import dynamic from 'next/dynamic';
import { useMediaQuery, useWL } from '../../hooks';
import { DevRevSvg, MenuburgerSvg, ClockSvg } from '../../components/svgs';
import { LazyRender, LoadingBar } from '../../components';
import { useRouter } from 'next/router';
import { Timer } from '../../utils';
import WatchLaterDrop from './WatchLaterDrop';

const MenuMain = dynamic(() => import('./Menu/MenuMain'), {
  // eslint-disable-next-line react/display-name
  loading: () => <MenuPlaceHolder />,
  ssr: false
});

const MobileMenu = dynamic(() => import('./MobileMenu'), {
  // eslint-disable-next-line react/display-name
  loading: () => <div style={{ display: 'none' }}></div>,
  ssr: false
});

const initialState = {
  showPlaylistMenu: false,
  showCommunityMenu: false,
  ShowWatchLater: false,
  ShowMenu: false
};

function reducer(state, action) {
  switch (action.type) {
    case 'playlist':
      return {
        showPlaylistMenu: true,
        showCommunityMenu: false,
        ShowWatchLater: false,
        ShowMenu: false
      };
    case 'community':
      return {
        showPlaylistMenu: false,
        showCommunityMenu: true,
        ShowWatchLater: false,
        ShowMenu: false
      };
    case 'menu':
      return {
        showPlaylistMenu: false,
        showCommunityMenu: false,
        ShowWatchLater: false,
        ShowMenu: true
      };
    case 'wl':
      return {
        showPlaylistMenu: false,
        showCommunityMenu: false,
        ShowWatchLater: true,
        ShowMenu: false
      };
    case 'reset':
      return {
        showPlaylistMenu: false,
        showCommunityMenu: false,
        ShowWatchLater: false,
        ShowMenu: false
      };
    default:
      return state;
  }
}

const Navigation = () => {
  const OpenedMenuCache = useRef(false);
  const innerWidthCache = useRef(false);
  const WLCartInterval = useRef(null);
  const Router = useRouter();

  const [MenuState, dispatchMenu] = useReducer(reducer, initialState);
  const {
    ShowMenu,
    showPlaylistMenu,
    showCommunityMenu,
    ShowWatchLater
  } = MenuState;
  const [disableTransition, setDisableTransition] = useState(false);
  const [WlPulseTrigger, setWlPulseTrigger] = useState(false);

  const mediaQueryMatches = useMediaQuery('max-width', 735);
  const [StoredValue] = useWL();

  const show = useMemo(() => showPlaylistMenu || showCommunityMenu, [
    showCommunityMenu,
    showPlaylistMenu
  ]); // || more dropMenus

  if (typeof window !== 'undefined')
    innerWidthCache.current = window.innerWidth >= 735;

  const isMobileView = !(!mediaQueryMatches && innerWidthCache.current);
  const IsWatchLater = StoredValue?.wl.length > 0;

  OpenedMenuCache.current = ShowMenu || show || ShowWatchLater;

  useEffect(() => {
    if (!mediaQueryMatches && ShowMenu) {
      dispatchMenu({ type: 'reset' });
    }
  }, [mediaQueryMatches]);

  useEffect(() => {
    if (IsWatchLater) {
      setWlPulseTrigger(true);
      Timer(500).then(() => {
        setWlPulseTrigger(false);
      });
    }
  }, [StoredValue]);

  const ResetMenuOnRouteChange = () => {
    // event callback cannot access the latest state.
    // Therefore, we can incorporate useRef to solve this problem.
    if (OpenedMenuCache.current) {
      dispatchMenu({ type: 'reset' });
      let PlaylistNode = document.getElementById('playlist-menu');
      let CommunityNode = document.getElementById('community-menu');
      // Add every new dropdown on navbar here.
      if (PlaylistNode?.style.display === 'block')
        PlaylistNode.style.display = 'none';
      if (CommunityNode?.style.display === 'block')
        CommunityNode.style.display = 'none';
    }
  };

  Router?.events?.on('routeChangeComplete', ResetMenuOnRouteChange);
  Router?.events?.on('routeChangeError', ResetMenuOnRouteChange);

  const HandleDisableTransition = () => {
    setDisableTransition(true);
  };

  const HandleEnableTransition = () => {
    setDisableTransition(false);
  };

  const HandleUserTabAction = () => {
    const TabClass = 'user-is-tabbing';
    const BodyClass = document.body.classList;
    if (BodyClass.contains(TabClass)) BodyClass.remove(TabClass);
  };

  const HandleMenu = useCallback(() => {
    HandleUserTabAction();
    dispatchMenu({ type: ShowMenu ? 'reset' : 'menu' });
  }, [ShowMenu]);

  const HandleKeyPressMenu = (e) => {
    if (e.key === 'Enter' || e.which === 13) {
      dispatchMenu({ type: ShowMenu ? 'reset' : 'menu' });
    }
  };

  const WlCartEvent = (e) => {
    const target = e.target;
    const WLCart = document.getElementById('watch-later-cart');
    const WLClock = document.getElementById('wl-clock');

    if (!WLCart?.contains(target) && !WLClock.contains(target)) {
      clearInterval(WLCartInterval.current);
      dispatchMenu({ type: 'reset' });
      document.removeEventListener('click', WlCartEvent);
    }
  };

  const HandleWatchLaterCart = useCallback(() => {
    HandleUserTabAction();
    if (ShowWatchLater) {
      dispatchMenu({ type: 'reset' });
    } else {
      dispatchMenu({ type: 'wl' });
      document.addEventListener('click', WlCartEvent);
    }
  }, [ShowWatchLater]);

  const HandleKeyPressWlCart = (e) => {
    if (e.key === 'Enter' || e.which === 13) HandleWatchLaterCart();
  };

  const WlCartMouseIn = useCallback(() => {
    if (WLCartInterval.current) clearInterval(WLCartInterval.current);
  }, [WLCartInterval.current]);

  const WlCartMouseLeave = useCallback(() => {
    if (ShowWatchLater) {
      WLCartInterval.current = setInterval(() => {
        clearInterval(WLCartInterval.current);
        dispatchMenu({ type: 'reset' });
      }, 900);
    }
  }, [WLCartInterval.current, ShowWatchLater]);

  return (
    <Nav menuIsOpen={ShowMenu}>
      <LoadingBar></LoadingBar>
      <NavContainer>
        <MenuContainer
          onMouseEnter={HandleDisableTransition}
          onMouseLeave={HandleEnableTransition}
        >
          <NavLogo data-content="Home">
            <Link href="/">
              <a aria-label="home">
                <DevRevSvg />
              </a>
            </Link>
          </NavLogo>
          <LazyRender render={!isMobileView} Suspense={<MenuPlaceHolder />}>
            <MenuMain
              disTran={disableTransition}
              Show={show}
              dispatchMenu={dispatchMenu}
              wlOpened={ShowWatchLater}
            />
          </LazyRender>
        </MenuContainer>
        <RightContainer>
          <WatchLaterContainer
            onClick={HandleWatchLaterCart}
            id="wl-clock"
            aria-label="watch later"
            tabIndex={0}
            onKeyPress={HandleKeyPressWlCart}
            onMouseEnter={WlCartMouseIn}
            onMouseLeave={WlCartMouseLeave}
          >
            <ClockIconContainer>
              <ClockSvg />
            </ClockIconContainer>
            <WLCounter as="span" Trigger={WlPulseTrigger}>
              <span>{StoredValue?.wl.length ?? 0}</span>
            </WLCounter>
            <WLMenuTriangle
              style={{ bottom: '-3px', right: '20%' }}
              Enable={ShowWatchLater}
            ></WLMenuTriangle>
          </WatchLaterContainer>
          <WatchLaterDrop
            Show={ShowWatchLater}
            ShoppingCartMouseIn={WlCartMouseIn}
            ShoppingCartMouseLeave={WlCartMouseLeave}
          />
          <MenuBurgerContainer
            onClick={HandleMenu}
            onKeyPress={HandleKeyPressMenu}
          >
            <Burger>
              <MenuburgerSvg menuIsOpen={ShowMenu} />
            </Burger>
            <BurgerTxt tabIndex={0}>menu</BurgerTxt>
          </MenuBurgerContainer>
          <LazyRender render={isMobileView} Suspense={null}>
            <MobileMenu ShowMenu={ShowMenu} />
          </LazyRender>
        </RightContainer>
      </NavContainer>
    </Nav>
  );
};

const MenuPlaceHolder = () => {
  return (
    <MenuMainPlaceholder>
      <div></div>
      <div></div>
      <div></div>
    </MenuMainPlaceholder>
  );
};

export default memo(Navigation);
