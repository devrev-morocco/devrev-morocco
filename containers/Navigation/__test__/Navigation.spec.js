/* eslint-disable no-undef */

// This will mock `next/dynamic`, so make sure to import it before any of your
// components.

import preloadAll from 'jest-next-dynamic';
import '../../../__test__/mock/matchMedia.mock'; // Must be imported before the tested file
import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Navigation from '..';

jest.mock('next/router', () => ({
  useRouter() {
    return {
      basePath: '',
      pathname: '/',
      route: '/',
      asPath: '/',
      query: {},
      push: jest.fn(),
      replace: jest.fn(),
      reload: jest.fn(),
      back: jest.fn(),
      prefetch: jest.fn(),
      beforePopState: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn()
      },
      isFallback: false
    };
  }
}));

const StoredValue = { wl: [] };
const setLocalStorage = () => void 0;

jest.mock('../../../hooks/useWL', () => ({
  useWL: () => [StoredValue, setLocalStorage]
}));

beforeAll(async () => {
  await preloadAll();
});

afterEach(cleanup);

test('render Navigation:', async () => {
  await preloadAll();
  render(<Navigation />);
  screen.debug();
  // expect(screen.getByText('Store')).toBeInTheDocument();
});

// If you want to mock the hook with different values in different tests:
// import * as hooks from '../../../hooks/useWL'
// jest.spyOn(hooks, 'useWL').mockImplementation(() => ([{wl:[(custom list)],()=>{}}));
