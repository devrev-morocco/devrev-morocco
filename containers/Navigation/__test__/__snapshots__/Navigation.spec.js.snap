// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Navigation: renders correctly 1`] = `
<nav
  className="styles__Nav-sc-5qfxzf-0 hZDOEY"
>
  <div
    className="styles__DisNone-k3o4wy-7 styles__LoadingBarContainer-lb27hq-3 bvobAG kZxxwL"
  />
  <div
    className="styles__DisFlex_AIC-k3o4wy-3 styles__NavContainer-sc-5qfxzf-1 uChuK dTTkhj"
  >
    <div
      className="styles__DisFlex_AIC-k3o4wy-3 styles__MenuContainer-sc-5qfxzf-3 uChuK esqTyS"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="styles__DisFlex_AIC_JCC-k3o4wy-5 styles__NavLogo-sc-5qfxzf-2 bqBaKh kVLPeR"
        data-content="Home"
      >
        <a
          aria-label="home"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            fill="#000"
            height={40}
            viewBox="0 0 500 500"
            width={40}
          >
            <path
              d="M35 1.9c-8.1 2.6-13.7 6.2-20.6 13.3C7.1 22.8 4.4 27 1.9 35.3.1 41.3 0 48.4 0 250s.1 208.7 1.9 214.7c2.5 8.3 5.2 12.5 12.5 20.1 7.1 7.3 12.5 10.8 21 13.3 5.8 1.8 14.7 1.9 214.6 1.9s208.8-.1 214.6-1.9c8.5-2.5 13.9-6 21-13.3 7.3-7.6 10-11.8 12.5-20.1 1.8-6 1.9-13.1 1.9-214.7 0-199.9-.1-208.8-1.9-214.6-2.5-8.5-6-13.9-13.3-21-7.6-7.3-11.8-10-20.1-12.5C458.7.1 451.8 0 249.5.1 53.6.2 40.2.3 35 1.9zM384 250v134H116V116h268v134z"
            />
            <path
              d="M124 250v126h252V124H124v126zm67.5-82.5c10.9 2.8 19.5 10.6 23.6 21.3 1.7 4.6 2 7.3 1.7 14.2-.3 7.3-.9 9.4-3.8 14.8-3.9 7.4-8.3 11.5-16 15.1-5 2.4-6.9 2.6-20.2 2.9l-14.8.4V166h11.8c7.3 0 14 .6 17.7 1.5zm76.3 5.7l.3 6.8H243v14h24v14h-24v14h25v14h-41v-70l20.3.2 20.2.3.3 6.7zm29.2 9.5c3.4 9.3 7.3 20.3 8.8 24.5 1.5 4.3 2.9 7.8 3.2 7.8.3 0 4.5-11 9.4-24.5l8.9-24.5H344l-1.3 3.2c-.7 1.8-7.1 17.6-14.2 35L315.7 236h-13.3l-13.3-32.9c-7.3-18.1-13.5-33.9-13.8-35-.5-2-.1-2.1 7.5-2.1h8l6.2 16.7zm-99.6 86.4c18.6 4.7 22 32.6 5.1 41.2l-4.6 2.3 9.6 12.2c5.2 6.7 9.5 12.4 9.5 12.7 0 .3-4.2.5-9.4.5h-9.3l-8.1-12.3-8.2-12.2V338h-16v-70h13.6c7.4 0 15.5.5 17.8 1.1zM263 275v7h-25v14h24v14h-24v14h25v14h-40v-70h40v7zm32.5 18.3c8.1 22.4 9.1 24.5 10 22.1.6-1.6 4.6-12.7 9.1-24.9l8.1-22 8.2-.3c4.4-.2 8.1.1 8.1.5s-6.2 16.1-13.8 34.8l-13.8 34h-13.6l-13.9-34.4c-7.6-18.9-13.9-34.6-13.9-34.8 0-.2 3.7-.3 8.3-.1l8.2.3 9 24.8z"
            />
            <path
              d="M177.7 180.7c-.4.3-.7 9.8-.7 21V222h3.3c9.9 0 18.6-6.6 20.5-15.5 2.7-12.6-3.9-23.2-15.8-25.5-6.2-1.2-6.5-1.2-7.3-.3zM182 290.4V301h4.4c6.1 0 9.2-1.4 11.1-4.9 2.3-4.6 1.9-8.6-1.3-12.2-2.4-2.7-3.7-3.3-8.5-3.7l-5.7-.4v10.6z"
            />
          </svg>
        </a>
      </div>
      <div
        className="styles__DisFlex_AIC_JCC-k3o4wy-5 styles__MenuMainPlaceholder-sc-5qfxzf-4 bqBaKh dSSjnX"
      >
        <div />
        <div />
        <div />
      </div>
    </div>
    <div
      className="styles__DisFlex_AIC-k3o4wy-3 styles__RightContainer-sc-5qfxzf-8 uChuK cmvUfb"
    >
      <div
        aria-label="watch later"
        className="styles__DisFlex_AIC_JCC-k3o4wy-5 styles__WatchLaterContainer-sc-5qfxzf-9 bqBaKh dTjgCq"
        id="wl-clock"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        tabIndex={0}
      >
        <div
          className="styles__ClockIconContainer-sc-5qfxzf-10 ecfSrv"
        >
          <svg
            viewBox="0 0 24 24"
          >
            <path
              d="M12 3.67c-4.58 0-8.33 3.75-8.33 8.33s3.75 8.33 8.33 8.33 8.33-3.75 8.33-8.33S16.58 3.67 12 3.67zm3.5 11.83l-4.33-2.67v-5h1.25v4.34l3.75 2.25-.67 1.08z"
              fill="#fff"
            />
          </svg>
        </div>
        <span
          className="styles__AbsolutePosition-k3o4wy-1 styles__CounterContainer-sc-5qfxzf-11 styles__WLCounter-sc-5qfxzf-12 ezonLg ldNrNC"
        >
          <span>
            0
          </span>
        </span>
        <div
          className="styles__AbsolutePosition-k3o4wy-1 StyledComponents__Triangle-fvydi3-0 StyledComponents__WLMenuTriangle-fvydi3-2 ezonLg BijnC bOkNdc"
          style={
            Object {
              "bottom": "-3px",
              "right": "20%",
            }
          }
        />
      </div>
      <div
        className="styles__AbsolutePosition-k3o4wy-1 style__Cart_Container-bkbm6i-0 style__WLCart_Container-bkbm6i-1 ezonLg cnDttn gikETA"
        id="watch-later-cart"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="styles__DisFlex_AIC-k3o4wy-3 style__WLCartHeader-bkbm6i-3 uChuK kYdqXW"
        />
        <div
          className="styles__RelativePosition-k3o4wy-0 style__WLCartContainerWrap-bkbm6i-2 hxbOxp bGbMJc"
        >
          <div
            className="styles__RelativePosition-k3o4wy-0 style__WlNoResultsContainer-bkbm6i-7 hxbOxp cZDaUw"
          >
            <div
              className="styles__DisFlex_AIC_JCC-k3o4wy-5 style__WlNoResultsWrapper-bkbm6i-8 bqBaKh kiNCbo"
            >
              <div
                className="style__WlNoResultsHeader-bkbm6i-9 KpNwu"
              >
                Your watch later queue is empty
              </div>
              <div
                className="separator"
                style={
                  Object {
                    "width": "50%",
                  }
                }
              />
              <p
                style={
                  Object {
                    "fontSize": "0.85em",
                    "margin": "10px 20px",
                  }
                }
              >
                Hover over a video thumbnail and then click on the “Watch Later” button (it has a Clock icon). The video will be instantly added to your Watch Later queue.
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        className="styles__DisNone-k3o4wy-7 styles__MenuBurgerContainer-sc-5qfxzf-5 bvobAG bsxRHe"
        onClick={[Function]}
        onKeyPress={[Function]}
      >
        <div
          className="styles__DisFlex_AIC_JCC-k3o4wy-5 styles__Burger-sc-5qfxzf-6 bqBaKh dqBrDv"
        >
          <div
            className="styles__RelativePosition-k3o4wy-0 styles__MenuburgerContainer-l71sm3-0 hxbOxp QtwKX"
          >
            <div
              className="styles__RelativePosition-k3o4wy-0 styles__XSpan-l71sm3-1 styles__BurgerXSpan-l71sm3-2 hxbOxp fwBaRB jmDvGZ"
            />
            <div
              className="styles__RelativePosition-k3o4wy-0 styles__XSpan-l71sm3-1 styles__BurgerXSpan-l71sm3-2 hxbOxp fwBaRB jmDvGZ"
            />
            <div
              className="styles__RelativePosition-k3o4wy-0 styles__XSpan-l71sm3-1 styles__BurgerXSpan-l71sm3-2 hxbOxp fwBaRB jmDvGZ"
            />
          </div>
        </div>
        <div
          className="styles__BurgerTxt-sc-5qfxzf-7 eXczMv"
          tabIndex={0}
        >
          menu
        </div>
      </div>
    </div>
  </div>
</nav>
`;
